src/charm.py

```diff
@@ -41,83 +65,51 @@ def __init__(self, *args):
 
     def _update_prometheus_jobs(self, _):
         """Setup Prometheus scrape configuration for external targets."""
-        self.unit.status = ActiveStatus()
-        if not self.unit.is_leader():
-            return
-
-        jobs = self._scrape_jobs()
-        if not jobs:
-            return
-
-        for relation in self.model.relations[self._prometheus_relation]:
-            relation.data[self.app]["scrape_jobs"] = json.dumps(jobs)
+        if self.unit.is_leader():
+            if jobs := self._scrape_jobs():
+                for relation in self.model.relations[self._prometheus_relation]:
+                    relation.data[self.app]["scrape_jobs"] = json.dumps(jobs)
 
-    def _scrape_jobs(self):
-        targets = self._targets()
+                self.unit.status = ActiveStatus()
 
-        jobs = (
-            [
+    def _scrape_jobs(self) -> list:
+        if targets := self._targets():
+            labels = self._labels()
+            return [
                 {
                     "job_name": self._job_name(),
                     "metrics_path": self.model.config["metrics-path"],
                     "static_configs": [
                         {
                             "targets": targets,
+                            "labels": labels,
                         }
                     ],
                 }
             ]
-            if targets
-            else []
-        )
-
-        labels = self._labels()
-        if jobs and labels:
-            jobs[0]["static_configs"][0]["labels"] = labels
 
-        return jobs
+        return []
 
-    def _targets(self):
-        if not (targets := self.model.config.get("targets", "")):
+    def _targets(self) -> list:
+        if not (config_targets := self.model.config.get("targets", "")):
+            self.unit.status = BlockedStatus("No targets specified")
             return []
 
-        urls = targets.split(",")
         targets = []
         invalid_targets = []
-        for url in urls:
-            if not (valid_address := self._validated_address(url)):
-                invalid_targets.append(url)
-                continue
-            targets.append(valid_address)
+        for config_target in config_targets.split(","):
+            if valid_address := _validated_address(config_target):
+                targets.append(valid_address)
+            else:
+                invalid_targets.append(config_target)
 
         if invalid_targets:
+            logger.error("Invalid targets found: %s", invalid_targets)
             self.unit.status = BlockedStatus(f"Invalid targets : {invalid_targets}")
+            return []
 
         return targets
 
-    def _validated_address(self, address):
```

Comment:
Why delete a whole function just to copy it elsewhere ?

Reply:
How does putting the method in module scope help in this case ? Is there any other object that will use this method ? Couldn't it be refactored to module scope in the future if such a use case arises ?